numba/targets/arraymath.py

```diff
@@ -999,3 +1019,192 @@ def dot_3(context, builder, sig, args):
+    check_lapack_return(context, builder, res)
+
+
+def call_xxgetri(context, builder, a_type, a_shapes, a_data, ipiv, work,
+                 lwork, info):
+    """
+    Call the LAPACK gettri function for the given argument.
+
+    This function computes the inverse of a matrix given its LU decomposition.
+
+    """
+    fnty = ir.FunctionType(ll_intc,
+                           [ll_char,                       # kind
+                            intp_t, ll_void_p, intp_t,     # n, a, lda
+                            ll_intc_p, ll_void_p,          # ipiv, work
+                            ll_intc_p, ll_intc_p           # lwork, info
+                           ])
+    fn = builder.module.get_or_insert_function(fnty, name="numba_xxgetri")
+
+    kind = get_blas_kind(a_type.dtype)
+    kind_val = ir.Constant(ll_char, ord(kind))
+
+    n = lda = a_shapes[0]
+
+    res = builder.call(fn, (kind_val, n,
+                            builder.bitcast(a_data, ll_void_p), lda,
+                            ipiv, builder.bitcast(work, ll_void_p),
+                            lwork, info
+                            ))
+    check_lapack_return(context, builder, res)
+
+
+def mat_inv(context, builder, sig, args):
+    """
+    Invert a matrix by first computing its LU decomposition.
+    """
+    xty = sig.args[0]
+    dtype = xty.dtype
+
+    x = make_array(xty)(context, builder, args[0])
+    x_shapes = cgutils.unpack_tuple(builder, x.shape)
+    m, n = x_shapes
+    check_c_int(context, builder, m)
+    check_c_int(context, builder, n)
+
+    # Allocate the return array (Numpy never works in place contrary to
+    # Scipy for which one can specify to whether or not to overwrite the
+    # input.
+    def create_out(a):
+        m, n= a.shape
+        if m != n:
+            raise ValueError("np.linalg.inv can only work on square "
+                             "arrays.")
+        return a.copy()
+
+    out = context.compile_internal(builder, create_out,
+                                   signature(sig.return_type, *sig.args), args)
+    o = make_array(xty)(context, builder, out)
+
+    # Allocate the array in which the pivot indices are stored.
+    ipiv_t = types.Array(types.intc, 1, 'C')
+    i = _empty_nd_impl(context, builder, ipiv_t, (m,))
+
+    info = cgutils.alloca_once(builder, ll_intc)
+
+    # Compute the LU decomposition of the matrix.
+    call_xxgetrf(context, builder, xty, x_shapes, o.data, i.data,
+                 info)
+
+    zero = ir.Constant(ll_intc, 0)
+    info_val = builder.load(info)
+    lapack_error = builder.icmp_signed('!=', info_val, zero)
+    invalid_arg = builder.icmp_signed('<', info_val, zero)
+
+    with builder.if_then(lapack_error, False):
+        with builder.if_else(invalid_arg) as (then, otherwise):
+            raise_err = context.call_conv.return_user_exc
+            with then:
+                raise_err(builder, ValueError,
+                          ('One argument passed to getrf is invalid',)
```

Comment:
@pitrou would being able to raise with an interpolated string be possible following work on https://github.com/numba/numba/issues/1670 ? It'd be nice in cases like this to say "Argument X is incorrect", and similarly below, if the matrix is singular, it'd be nice to be able to say where and why.